text/0456-elasticache-l2.md

```diff
@@ -0,0 +1,359 @@
+# Amazon ElastiCache L2
+
+* **Original Author(s):**: @dbartholomae
+* **Tracking Issue**: #456
+* **API Bar Raiser**: @corymhall
+
+The `aws-elasticache` construct library allows you to create Amazon ElastiCache 
+data stores with just a few lines of code. As with most construct libraries,
+you can also easily define permissions and metrics using a simple API.
+
+The amount of effort needed to create these resources is about the same as doing
+it using the AWS console, sometimes even less due to easier set up of VPC connectivity.
+
+## Working Backwards
+
+### CHANGELOG
+
+`feat(elasticache): RedisReplicationGroup L2`
+
+### README
+
+---
+
+# Amazon ElastiCache Construct Library
+
+[Amazon ElastiCache](https://aws.amazon.com/elasticache/) is a fully managed,
+in-memory caching service supporting flexible, real-time use cases. You can use
+ElastiCache for caching, which accelerates application and database performance,
+or as a primary data store for use cases that don't require durability like
+session stores, gaming leaderboards, streaming, and analytics. ElastiCache is
+compatible with Redis and Memcached.
+
+This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk)
+project. It allows you to define ElastiCache Redis replication groups.
+
+## Constructs
+
+### Defining a RedisReplicationGroup
```

Comment:
On this, I am missing deep enough understanding of ElastiCache to make the best trade-offs. E.g. for `UseOnlineResharding`, there is [a warning that you should update `NumNodeGroups` and `NodeGroupConfiguration` only in isolation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html#cfn-attributes-updatepolicy-useonlineresharding), which might not be obvious when using AWS CDK.

I agree on making ids required and will add it to the RFC. Seems to be a no-brainer.

The last point also speaks against setting `UseOnlineResharding` by default.

After looking into this, I would actually suggest that we make `useOnlineResharding` a required property for now. This way, introducing a default value in a later PR is a non-breaking change that can be done with close support from someone with deep ElastiCache understanding.
If you actually have deep enough ElastiCache understanding (or can bring in an expert who has), I'm also open to just implementing what you discussed. Though I would be prefer to not delay the implementation too much based on this.

Reply:
Just saw that `NodeGroupConfiguration` currently is marked as a prop that is not part of the first implementation anyways, so  I would only add the information on making the id required if we actually go deeper into this part anyways. Otherwise, I would leave this up for a future RFC or implementation discussion.

Reply:
The default behavior should be to use online resharding. It increases availability and allows you to easily add and remove shards from your cluster. Offline resharding is necessary when you want to do more than just add or remove shards.

Reply:
@dorser if this is enabled by default and someone who doesn't know about the details updates NumNodeGroups and other properties at the same time, what would happen? The docs caution against this, and I would aim for an L2 construct to be usable without deep understanding of the service details.
Similarly, would we need to handle `NodeGroupConfiguration.Slots` differently in that case to align with the recommended approach of deleting it from the template after initial deployment?

Reply:
I'm not sure how to proceed on this discussion. @dorser @corymhall what do you think?

Reply:
I would suggest that by default you omit the node group configuration slots and you always set the UseOnlineResharding. This is a bit of legacy about our service, but we initially launched only with `OfflineResharding`. Customers cared way more about online resharding than they did about configuring slots, so when we added online resharding and we didn't allow custom slot configurations. If someone **REALLY** wants set set slots, they can use the L1 constructs, but my guess is that is a very small number of users.

Reply:
@dbartholomae based on @madolson let's have `UseOnlineResharding` as the default.